### PR TITLE
Make LeanDefinitionalEngine thread safe (#227)

### PR DESCRIPTION
**Pull request by andrewmwells-amazon**
*Issue #, if available:*
#227 

*Description of changes:*
Make `LeanDefinitionalEngine` thread safe.

### CODE REVIEW

**Review by khieta — APPROVED** [review]
Nice! Likely want to wait until #366 is merged so the CI is doing its job properly.

**Review by shaobo-he-aws — APPROVED** [review]

**Review by cdisselkoen — APPROVED** [review]

**Pull request merged**
